core/shared/src/main/scala/zio/Chunk.scala

```diff
@@ -1117,6 +1160,136 @@ object Chunk {
     override def toArray[A1 >: A](n: Int, dest: Array[A1]): Unit = { val _ = vector.copyToArray(dest, n, length) }
   }
 
+  private[zio] object Buff {
+    final def arrayBacked[A](buffer: Buffer): Option[Chunk[A]] =
+      if (!buffer.hasArray()) None
+      else {
+        val pos = buffer.position()
+        val lim = buffer.limit()
+        val cap = buffer.capacity()
+        val arr = buffer.array().asInstanceOf[Array[A]]
+        val off = buffer.arrayOffset()
+
+        val chunk = if (off == 0 && pos == 0 && cap == arr.length && lim == cap) {
+          // best case, reuse the underlying array
+          fromArray(arr)
+        } else {
+          implicit val classTag: ClassTag[A] = ClassTag(arr.getClass.getComponentType)
+          val len                            = lim - pos
+          val dest                           = Array.ofDim[A](len)
```

Comment:
Rather than re-implement the slicing logic, let's allocate an array of size `remaining()`, and use `Buffer#get(arr)` to copy the data into the array. Also, to avoid mutating the position, use `mark()` before `get` and `reset` afterwards.